### PR TITLE
Add compaction checker to check whether compaction should be scheduled

### LINKED ISSUE #4871: add compaction cruiser for event-based compaction framework
## Feature request

**Is your feature request related to a problem? Please describe.**

each tablet needs to be base-compacted when the time elapse from last base compaction is large than config::base_compaction_interval_seconds_since_last_operation. But in event-based compaction framework, base compaction will not be scheduled if the tablets do not have enough compaction score.

**Describe the solution you'd like**

add a compaction cruiser in the background thread to check whether should schedule base compaction for each tablet. This check can be scheduled by interval and should control the step of each check to limit the cost.

### PR DESCRIPTION
**Pull request by ABingHuang**
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #4871

## Problem Summary(Required) ：

Base compaction will be scheduled every day. So add compaction checker to check
whether base compaction should be scheduled. It will be scheduled every 30 minutes
after last run and check 1000 tablets every seconds, which makes sure that compaction checker
not to hurt system performance.

### CODE REVIEW

File: be/src/storage/storage_engine.cpp [hunk 1]
```diff
@@ -543,6 +544,43 @@ void StorageEngine::_start_clean_fd_cache() {
     VLOG(10) << "Cleaned file descriptor cache";
 }
 
+void StorageEngine::compaction_check() {
+    int checker_one_round_sleep_time_s = 1800;
+    while (true) {
```
Comment (meegoo) [inline be/src/storage/storage_engine.cpp:549]:
```suggestion
    while (!_bg_worker_stopped.load(std::memory_order_consume)) {
```

File: be/src/storage/storage_engine.cpp [hunk 2]
```diff
@@ -543,6 +544,43 @@ void StorageEngine::_start_clean_fd_cache() {
...
+        bool finished = _tablet_manager->get_next_batch_tablets(batch_size, &tablets);
+        for (auto& tablet : tablets) {
+            _compaction_manager->update_tablet(tablet, true, false);
+        }
+        tablets_num_checked += tablets.size();
+        tablets.clear();
+        if (finished) {
+            break;
+        }
+        std::this_thread::sleep_for(std::chrono::milliseconds(batch_sleep_time_ms));
```
Comment (meegoo) [inline be/src/storage/storage_engine.cpp:579]: one second sleep seems not enough

**Review by meegoo — CHANGES_REQUESTED** [review]

File: be/src/storage/storage_engine.cpp [hunk 3]
```diff
@@ -543,6 +544,43 @@ void StorageEngine::_start_clean_fd_cache() {
...
+    int checker_one_round_sleep_time_s = 1800;
+    while (true) {
+        MonotonicStopWatch stop_watch;
+        stop_watch.start();
+        LOG(INFO) << "start to check compaction";
+        size_t num = compaction_check_one_round();
+        stop_watch.stop();
+        LOG(INFO) << num << " tablets checked. time elapse:" << stop_watch.elapsed_time() / 1e9 << " seconds."
+                  << " compaction checker will be scheduled again in " << checker_one_round_sleep_time_s << " seconds";
+        std::this_thread::sleep_for(std::chrono::seconds(checker_one_round_sleep_time_s));
```
Comment (chaoyli) [inline be/src/storage/storage_engine.cpp:557]: This will make the exit gracefully stall on 1800 seconds.

**Review by chaoyli — CHANGES_REQUESTED** [review]

File: be/src/storage/storage_engine.cpp (hunk 2 above)
Comment (ABingHuang) [reply be/src/storage/storage_engine.cpp:579]: it is used to control the speed of checker, and do not use cpu that much. It does not limit the candidate queue size.

File: be/src/storage/storage_engine.cpp (hunk 3 above)
Comment (ABingHuang) [reply be/src/storage/storage_engine.cpp:557]: fixed

File: be/src/storage/olap_server.cpp [hunk 4]
```diff
@@ -122,6 +122,11 @@ Status StorageEngine::start_bg_threads() {
             compaction_scheduler.schedule();
         });
         Thread::set_thread_name(_compaction_scheduler, "compact_sched");
+        LOG(INFO) << "compaction scheduler started";
```
Comment (chaoyli) [inline be/src/storage/olap_server.cpp:125]:
```
LOG(INFO) << "compaction scheduler started";
```
The log is non-sense.

File: be/src/storage/storage_engine.cpp [hunk 5]
```diff
@@ -543,6 +549,47 @@ void StorageEngine::_start_clean_fd_cache() {
     VLOG(10) << "Cleaned file descriptor cache";
 }
 
+void StorageEngine::compaction_check() {
+    int checker_one_round_sleep_time_s = 1800;
+    while (!bg_worker_stopped()) {
+        MonotonicStopWatch stop_watch;
+        stop_watch.start();
+        LOG(INFO) << "start to check compaction";
```
Comment (chaoyli) [inline be/src/storage/storage_engine.cpp:557]:
```
LOG(INFO) << "start to check compaction";
```
The log is non-sense.

Comment (chaoyli) [inline be/src/storage/storage_engine.cpp:557]:
```
LOG(INFO) << "start to check compaction";
```
The log is non-sense.

**Review by chaoyli — CHANGES_REQUESTED** [review]

File: be/src/storage/storage_engine.cpp (hunk 5 above)
Comment (ABingHuang) [reply be/src/storage/storage_engine.cpp:557]: it is useful to get background task info. I tend to keep it.

File: be/src/storage/olap_server.cpp (hunk 4 above)
Comment (ABingHuang) [reply be/src/storage/olap_server.cpp:125]: it can help to locate the background task start progress. I tend to keep it.

**Review by chaoyli — APPROVED** [review]

**Review by decster — APPROVED** [review]

**Pull request merged**